### PR TITLE
[passport] Strongly type serialize-/deserializeUser with Express.User

### PR DESCRIPTION
**Pull request by HoldYourWaffle**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
* [deserializeUser](https://github.com/jaredhanson/passport/blob/1fd591d5cdcf6516d9eb446216843223c0e020c4/lib/strategies/session.js#L60-L67)
* [serializeUser](https://github.com/jaredhanson/passport/blob/1fd591d5cdcf6516d9eb446216843223c0e020c4/lib/sessionmanager.js#L14-L25)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

This is of course a breaking change, but as far as I can see the current situation (introduced in #12446) does not make sense (anymore). @RandScullard could you shine some light on your rationale for these generics?

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## passport-local-mongoose ([<kbd>unpkg</kbd>](https://unpkg.com/browse/passport-local-mongoose@latest/))
was missing the following properties:
1. errors
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 9b8ddb6305d56f003318b542398bff5b00404b74
</p>

**Comment by RandScullard** [comment]
@HoldYourWaffle Can you explain what you are trying to accomplish here? The reason I added the TUser generic parameter was to get strong typing against whatever custom user object you are using with passport. If I'm understanding your change correctly, you are proposing to remove TUser and use Express.User instead. But Express.User appears to be declared as an empty interface, so I'm not sure how that would help anything. Am I missing something?

**Comment by HoldYourWaffle** [comment]
`Express.User` is meant to be extended using [module augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation).
Most notably, the `Request.user` property used by passportjs has this type:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/8e7a71356a0bdb45a15261b4219ae9b34724562e/types/passport/index.d.ts#L22-L24
Since the values that I have changed the types of ultimately end up in `Request.user` (see the links I provided), they should also have the same type.

Does this answer your question?

**Comment by RandScullard** [comment]
OK, thanks for the clarification. Now I understand the context around introducing the Express.User type. I agree that for consistency Express.User should be used in serializeUser and deserializeUser, and that the generic parameter is no longer needed. Too bad it's a breaking change, but I guess the needed changes will go hand in hand with setting up the module augmentation for the User type.

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by HoldYourWaffle** [comment]
Ready to merge

**Comment by marcobiedermann** [comment]
I'm sorry to ask but could you give some further examples on how to use it.
In my setup, I get `Property 'id' does not exist on type 'User'.`

```ts
passport.serializeUser<string>((user, done): void => {
  done(null, user.id);
});
```

Because of the Generic the compiler knows that it's of type `string` but `Express.User` is defined as an empty object.

Is there any way to declare it without using `global` as shown in the examples?

File: types/passport-google-oauth20/index.d.ts [hunk 1]
```diff
@@ -41,7 +41,7 @@ export interface Profile extends passport.Profile {
     _json: any;
 }
 
-export type VerifyCallback = (err?: string | Error, user?: any, info?: any) => void;
+export type VerifyCallback = (err?: string | Error, user?: Express.User, info?: any) => void;
```
Comment (bodinsamuel) [inline types/passport-google-oauth20/index.d.ts:44]: I don't fully agree with this change that does not reflect the truth.
what you verify is the response from the oauth call and it's not yet an Express.User. What is an express.user is what is returned at the end of the VerifyCallback and can be dramatically different.
It's breaking for unnecessary reason I think

File: types/passport-oauth2/index.d.ts [hunk 2]
```diff
@@ -52,7 +52,7 @@ declare namespace OAuth2Strategy {
         verify(req: Request, state: string, meta: Metadata, callback: StateStoreVerifyCallback): void;
     }
 
-    type VerifyCallback = (err?: Error | null, user?: object, info?: object) => void;
+    type VerifyCallback = (err?: Error | null, user?: Express.User, info?: object) => void;
```
Comment (bodinsamuel) [inline types/passport-oauth2/index.d.ts:55]: I don't fully agree with this change that does not reflect the truth.
what you verify is the response from the oauth call and it's not yet an Express.User. What is an express.user is what is returned at the end of the VerifyCallback and can be dramatically different.
It's breaking for unnecessary reason I think

**Comment by sofusskovgaard** [comment]
@marcobiedermann You have to overwrite the declaration of the `Express.User` interface.. _I believe..._

I got it working by inserting the following anywhere in my code.

```typescript
declare global {
    namespace Express {
      interface User {
        id: number;
      }
    }
}
```

**Comment by ejose19** [comment]
This broke ```koa-passport``` as now it can't be typed in the function, ie:

```ts
passport.serializeUser((user: CustomUser, done) => {
  done(null, user);
});
```

errors with: 
```
Types of parameters 'user' and 'user' are incompatible.
   Type 'User' is missing the following properties from type 'CustomUser': id
```

**Comment by HoldYourWaffle** [comment]
@ejose19 I'm not familiar with `koa`, but the error suggests that either declaration merging for `CustomUser` was not set up correctly, or `koa-passport` is not using the correct `User` type in their definitions.

**Comment by DrJohnWagner** [comment]
Test [passport-local-mongoose-tests.ts](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/passport-local-mongoose/passport-local-mongoose-tests.ts) gives me an error on line 126 when `interface User` is redefined in the block:
```
type _User = User;

declare global {
    namespace Express {
        interface User extends PassportLocalModel<_User> {
        }
    }
}

```
The error I get is TS2300:
```
type Express.User = PassportLocalModel<User>
Duplicate identifier 'User'. ts(2300)
index.d.ts(20, 19): 'User' was also declared here
```
where Index.d.ts(20, 19) refers to node_modules/@types/passport/index.d.ts

My tsconfig.json:
```
{
  "compilerOptions": {
    "target": "es5",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "allowJs": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "react",
    "outDir": "./build-server/",
    "noFallthroughCasesInSwitch": true,
    "noEmit": true
  },
  "include": [
    "server/**/*"
  ],
  "exclude": [
    "node_modules"
  ]
}
```
My package.json:
```
{
  "name": "fullstack-app",
  "version": "0.1.0",
  "private": true,
  "proxy": "http://localhost:9000",
  "type": "module",
  "engines": {
    "node": ">=12.0.0",
    "yarn": ">=1.15.*"
  },
  "dependencies": {
    "axios": "^0.21.1",
    "bluebird": "^3.7.2",
    "compression": "^1.7.4",
    "connect-mongo": "^4.4.1",
    "cors": "^2.8.5",
    "dotenv": "^8.2.0",
    "errorhandler": "^1.5.1",
    "express": "^4.17.1",
    "express-session": "^1.17.1",
    "mongoose": "^5.9.15",
    "passport-local": "^1.0.0",
    "passport-local-mongoose": "^6.1.0",
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "react-scripts": "^4.0.3",
    "ts-loader": "^8.0.18",
    "web-vitals": "^1.1.1"
  },
  "devDependencies": {
    "@babel/core": "^7.13.10",
    "@types/bluebird": "^3.5.33",
    "@types/compression": "^1.7.0",
    "@types/cors": "^2.8.10",
    "@types/errorhandler": "^1.5.0",
    "@types/express": "^4.17.11",
    "@types/express-session": "^1.17.3",
    "@types/mongoose": "^5.10.4",
    "@types/node": "^14.0.5",
    "@types/passport-local": "^1.0.33",
    "@types/passport-local-mongoose": "^4.0.13",
    "@types/react": "^16.9.35",
    "@types/react-dom": "^16.9.8",
    "@typescript-eslint/eslint-plugin": "^4.19.0",
    "@typescript-eslint/parser": "^4.19.0",
    "axios-mock-adapter": "^1.18.1",
    "babel-plugin-module-resolver": "^4.1.0",
    "chalk": "^4.0.0",
    "concurrently": "^6.0.0",
    "dotenv-cli": "^4.0.0",
    "eslint": "^7.22.0",
    "eslint-config-airbnb": "^18.2.1",
    "eslint-config-prettier": "^8.1.0",
    "eslint-import-resolver-babel-module": "^5.2.0",
    "eslint-plugin-import": "^2.22.1",
    "eslint-plugin-jsx-a11y": "^6.4.1",
    "eslint-plugin-prettier": "^3.3.1",
    "eslint-plugin-react": "^7.23.0",
    "eslint-plugin-react-hooks": "^4.2.0",
    "mongodb-memory-server": "^6.6.1",
    "nodemon": "^2.0.4",
    "prettier": "^2.2.1",
    "rimraf": "^3.0.2",
    "typescript": "~3.9.3",
    "webpack": "^5.27.2"
  },
  "scripts": {
    "start-client": "react-scripts start",
    "build-client": "react-scripts build",
    "start-server": "concurrently -k \"yarn run watch-server\" \"yarn run renew-server\"",
    "build-server": "rimraf build-server && tsc --noEmit false",
    "watch-server": "tsc --watch --noEmit false",
    "renew-server": "nodemon -q --experimental-modules --es-module-specifier-resolution=node build-server/server.js",
    "serve": "cd build-server && node server.js",
    "seed": "babel-node tools/seedDatabase.js"
  },
  "eslintConfig": {
    "extends": "react-app"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}
```

**Pull request merged**